rs/xnet/payload_builder/src/lib.rs

```diff
@@ -1249,68 +1251,16 @@ impl PoolRefillTask {
 
     /// Queries all subnets for new slices and puts / appends them to the pool after
     /// validation against the given registry version.
-    async fn refill_pool(
-        &self,
-        pool_byte_size_soft_cap: usize,
-        slice_byte_size_max: usize,
-        registry_version: RegistryVersion,
-    ) {
-        let pool_slice_stats = {
-            let pool = self.pool.lock().unwrap();
-
-            if pool.byte_size() > pool_byte_size_soft_cap {
-                // Abort if pool is already full.
-                return;
-            }
-
-            pool.peers()
-                // Skip our own subnet, the loopback stream is routed separately.
-                .filter(|&&subnet_id| subnet_id != self.endpoint_resolver.subnet_id)
-                .map(|&subnet_id| (subnet_id, pool.slice_stats(subnet_id)))
-                .collect::<BTreeMap<_, _>>()
-        };
-
-        for (subnet_id, slice_stats) in pool_slice_stats {
-            let (stream_position, messages_begin, msg_count, byte_size) = match slice_stats {
-                // Have a cached stream position.
-                (Some(stream_position), messages_begin, msg_count, byte_size) => {
-                    (stream_position, messages_begin, msg_count, byte_size)
-                }
-
-                // No cached stream position, no pooling / refill necessary.
-                (None, _, _, _) => continue,
-            };
-
-            let (witness_begin, msg_begin, slice_byte_limit) = match messages_begin {
-                // Existing pooled stream, pull partial slice and append.
-                Some(messages_begin) if messages_begin == stream_position.message_index => (
-                    stream_position.message_index,
-                    stream_position.message_index + (msg_count as u64).into(),
-                    slice_byte_size_max.saturating_sub(byte_size),
-                ),
-
-                // No pooled stream, or pooled stream does not begin at cached stream position, pull
-                // complete slice from cached stream position.
-                _ => (
-                    stream_position.message_index,
-                    stream_position.message_index,
-                    slice_byte_size_max,
-                ),
-            };
-
-            if slice_byte_limit < SLICE_BYTE_SIZE_MIN {
-                // No more space left in the pool for this slice, bail out.
-                continue;
-            }
+    pub async fn refill_pool(&self, registry_version: RegistryVersion) {
+        let refill_stream_slice_indices = self.pool.lock().unwrap().refill_stream_slice_indices();
```

Comment:
Here the lock is a temporary variable that is released before we continue with the subsequent for loop.